GreedyParameterPointSampler.C

```diff
@@ -0,0 +1,1420 @@
+        }
+        else
+        {
+            d_iteration_started = false;
+            printErrorIndicatorToleranceNotMet();
+        }
+    }
+
+    // Precompute next error indicator point
+    // This will allow us to figure out if the greedy algorithm has terminated
+    // early without needing an extra call to the error indicator function.
+    getNextPointRequiringErrorIndicator();
+
+    return;
+}
+
+void
+GreedyParameterPointSampler::setConvergenceErrorIndicator(double proc_errors)
+{
+    printErrorIndicator(d_convergence_points[d_counter], proc_errors);
+
+    d_point_requiring_error_indicator_computed = false;
+
+    if (proc_errors > d_max_error)
+    {
+        d_max_error = proc_errors;
+    }
+
+    if (proc_errors >= d_error_indicator_tol)
+    {
+        d_iteration_started = false;
+        printErrorIndicatorToleranceNotMet();
+        getNextParameterPointAfterConvergenceFailure();
+        generateConvergenceSubset();
+    }
+    else
+    {
+        d_counter++;
+    }
+
+    if (d_counter == d_convergence_subset_size)
+    {
+        printConvergenceAchieved();
+        d_procedure_completed = true;
+    }
+
+    // Precompute next error indicator point
+    // This will allow us to figure out if the greedy algorithm has terminated
+    // early without needing an extra call to the error indicator function.
+    getNextPointRequiringErrorIndicator();
+}
+
+void
+GreedyParameterPointSampler::generateConvergenceSubset()
+{
+    d_convergence_points.clear();
+    d_convergence_points = generateRandomPoints(d_convergence_subset_size);
+}
+
+void
+GreedyParameterPointSampler::startConvergence()
+{
+    d_convergence_started = true;
+    d_max_error = 0;
+    d_counter = 0;
+    d_subset_counter = 0;
+
+    if (d_rank == 0)
+    {
+        std::string str;
+        str += "Error indicator tolerance " + std::to_string(d_error_indicator_tol) + " met. Computing convergence.\n";
+        agnosticPrint(str);
+    }
+
+    // Precompute next error indicator point
+    // This will allow us to figure out if the greedy algorithm has terminated
+    // early without needing an extra call to the error indicator function.
+    getNextPointRequiringErrorIndicator();
+}
+
+std::vector<Vector>
+GreedyParameterPointSampler::generateRandomPoints(int num_points)
+{
+    std::vector<Vector> random_points;
+
+    std::vector<std::uniform_real_distribution<double>> unif;
+    for (int i = 0; i < d_min_param_point.dim(); i++)
+    {
+        unif.push_back(std::uniform_real_distribution<double>(d_min_param_point.item(i), d_max_param_point.item(i)));
+    }
+
+    for (int i = 0; i < num_points; i++)
+    {
+        Vector point(d_min_param_point.dim(), false);
+        for (int j = 0; j < point.dim(); j++)
+        {
+            point.item(j) = unif[j](rng);
+        }
+        random_points.push_back(point);
+    }
+    return random_points;
+}
+
+std::shared_ptr<Vector>
+GreedyParameterPointSampler::getNearestROM(Vector point)
```

Comment:
It looks like some code duplication could be avoided by calling `getNearestROMIndex` from this function.

Reply:
They're slightly different to the point where it would be difficult to combine them, so I changed the function name and comment description in .h instead to better match what it is doing.